File: Languages/cpp/algorithms-dsa/leetcode/union-find/accounts_merge.cpp
/**
 * LeetCode 721: Accounts Merge
 * Topic: Union-Find
 * Difficulty: Medium
 *
 * Given accounts where each has a name and emails, merge accounts that
 * share at least one email. Union-Find on email ownership.
 * Time: O(n*k * α(n*k)) where k = avg emails per account, Space: O(n*k)
 */
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
#include <set>
#include <algorithm>
#include <cassert>
using namespace std;

class UnionFind {
    vector<int> parent;
public:
    UnionFind(int n) : parent(n) {
        for (int i = 0; i < n; i++) parent[i] = i;
    }
    int find(int x) {
        if (parent[x] != x) parent[x] = find(parent[x]);
        return parent[x];
    }
    void unite(int x, int y) {
        parent[find(x)] = find(y);
    }
};

vector<vector<string>> accountsMerge(vector<vector<string>>& accounts) {
    int n = accounts.size();
    UnionFind uf(n);
    unordered_map<string, int> emailToId;

    for (int i = 0; i < n; i++) {
        for (int j = 1; j < (int)accounts[i].size(); j++) {
            string& email = accounts[i][j];
            if (emailToId.count(email)) {
                uf.unite(i, emailToId[email]);
            } else {
                emailToId[email] = i;
            }
        }
    }

    unordered_map<int, set<string>> merged;
    for (int i = 0; i < n; i++) {
        int root = uf.find(i);
        for (int j = 1; j < (int)accounts[i].size(); j++)
            merged[root].insert(accounts[i][j]);
    }

    vector<vector<string>> res;
    for (auto& [root, emails] : merged) {
        vector<string> account = {accounts[root][0]};
        for (auto& e : emails) account.push_back(e);
        res.push_back(account);
    }
    return res;
}

int main() {
    vector<vector<string>> accounts = {
        {"John", "johnsmith@mail.com", "john_newyork@mail.com"},
        {"John", "johnsmith@mail.com", "john00@mail.com"},
        {"Mary", "mary@mail.com"},
        {"John", "johnnybravo@mail.com"}
    };
    auto res = accountsMerge(accounts);
    assert(res.size() == 3);

    // Find the merged John account (should have 3 emails)
    bool foundMerged = false;
    for (auto& acc : res) {
        if (acc.size() == 4 && acc[0] == "John") foundMerged = true;
    }
    assert(foundMerged);

    cout << "All tests passed!" << endl;
    return 0;
}

File: Languages/python/algorithms-dsa/leetcode/union-find/accounts_merge.py
"""
LeetCode #721 - Accounts Merge
Topic: Union-Find
Difficulty: Medium

Given a list of accounts where each account has a name and emails, merge
accounts that share at least one common email. Return merged accounts with
emails sorted.

Approach: Union-Find on email ownership. Map each email to an index, union
emails that belong to the same account, then group by root.

Time Complexity: O(n * k * α(n)) where k = avg emails per account
Space Complexity: O(n * k)
"""

from collections import defaultdict


class UnionFind:
    def __init__(self, n: int):
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, x: int) -> int:
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x: int, y: int) -> None:
        px, py = self.find(x), self.find(y)
        if px == py:
            return
        if self.rank[px] < self.rank[py]:
            px, py = py, px
        self.parent[py] = px
        if self.rank[px] == self.rank[py]:
            self.rank[px] += 1


def accounts_merge(accounts: list[list[str]]) -> list[list[str]]:
    email_to_id: dict[str, int] = {}
    email_to_name: dict[str, str] = {}
    idx = 0

    # Assign unique id to each email
    for account in accounts:
        name = account[0]
        for email in account[1:]:
            if email not in email_to_id:
                email_to_id[email] = idx
                idx += 1
            email_to_name[email] = name

    uf = UnionFind(idx)

    # Union all emails within same account
    for account in accounts:
        first_id = email_to_id[account[1]]
        for email in account[2:]:
            uf.union(first_id, email_to_id[email])

    # Group emails by root
    groups: dict[int, list[str]] = defaultdict(list)
    for email, eid in email_to_id.items():
        groups[uf.find(eid)].append(email)

    # Build result
    result = []
    for root, emails in groups.items():
        name = email_to_name[emails[0]]
        result.append([name] + sorted(emails))
    return result


if __name__ == "__main__":
    accounts = [
        ["John", "johnsmith@mail.com", "john_newyork@mail.com"],
        ["John", "johnsmith@mail.com", "john00@mail.com"],
        ["Mary", "mary@mail.com"],
        ["John", "johnnybravo@mail.com"],
    ]
    merged = accounts_merge(accounts)
    # Sort for comparison
    merged_sorted = sorted([sorted(a) for a in merged])
    expected = sorted([
        sorted(["John", "john00@mail.com", "john_newyork@mail.com", "johnsmith@mail.com"]),
        sorted(["John", "johnnybravo@mail.com"]),
        sorted(["Mary", "mary@mail.com"]),
    ])
    assert merged_sorted == expected

    # Single account
    assert accounts_merge([["Bob", "bob@mail.com"]]) == [["Bob", "bob@mail.com"]]

    print("All tests passed!")
